### PR TITLE
feat: typescript migration

### PR DESCRIPTION
**Pull request by jobo322**

### CODE REVIEW

File: src/classes/Gaussian.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ */
+export function widthToFWHM(width: number) {
+  return width * ROOT_2LN2;
+}
+
+/**
+ * Compute the value of width between the inflection points from Full Width at Half Maximum (FWHM).
+ * for more information check the [mathworld page](https://mathworld.wolfram.com/GaussianFunction.html)
+ * @param {number} fwhm - Full Width at Half Maximum.
+ * @returns {number} width
```
Comment (lpatiny) [inline src/classes/Gaussian.ts:125]: Also no need to specify type

File: src/classes/Gaussian2D.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
+import { GAUSSIAN_EXP_FACTOR } from '../util/constants';
+import erfinv from '../util/erfinv';
+
+import { widthToFWHM, fwhmToWidth } from './Gaussian';
+
+export { widthToFWHM, fwhmToWidth } from './Gaussian';
+
+export interface xyNumber {
```
Comment (lpatiny) [inline src/classes/Gaussian2D.ts:8]: Please use a PascalCase for interface names. In this case rather 'XYNumber'

File: src/classes/Gaussian2D.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+export function getSurface(options: GetVolumeOptions = {}) {
+  let { fwhm = 50, height = 1 } = options;
+
+  if (typeof fwhm !== 'object') fwhm = { x: fwhm, y: fwhm };
+
+  return (height * Math.PI * fwhm.y * fwhm.x) / Math.LN2 / 4;
+}
+
+function checkObject(input: number | xyNumber): xyNumber {
```
Comment (lpatiny) [inline src/classes/Gaussian2D.ts:216]: Please rename to ensureXYNumber()

File: package.json [hunk 4]
```diff
@@ -39,14 +57,17 @@
...
+    "eslint": "^7.32.0",
+    "eslint-config-cheminfo-typescript": "^8.0.9",
+    "jest": "^27.0.6",
+    "prettier": "^2.3.2",
+    "rimraf": "^3.0.2",
+    "ts-jest": "^27.0.4",
+    "typescript": "^4.3.5"
+  },
+  "dependencies": {
+    "jest-matcher-deep-close-to": "^2.0.1"
```
Comment (targos) [inline package.json:71]: devDependencies

File: src/classes/Gaussian2D.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
+import { GAUSSIAN_EXP_FACTOR } from '../util/constants';
+import erfinv from '../util/erfinv';
+
+import { widthToFWHM, fwhmToWidth } from './Gaussian';
+
+export { widthToFWHM, fwhmToWidth } from './Gaussian';
+
+export interface xyNumber {
+  [index: string]: number;
```
Comment (lpatiny) [inline src/classes/Gaussian2D.ts:9]: Why this line is there ?

File: src/util/getShapeGenerator.ts [hunk 6]
```diff
@@ -3,8 +3,15 @@ import { Gaussian2D } from '../classes/Gaussian2D';
 import { Lorentzian } from '../classes/Lorentzian';
 import { PseudoVoigt } from '../classes/PseudoVoigt';
 
-export function getShapeGenerator(options) {
-  let { kind = 'Gaussian', options: shapeOptions } = options;
+/**
+ * kind of shape
+ */
+type shapeKind = 'gaussian' | 'gaussian2D' | 'lorentzian' | 'pseudoVoigt';
```
Comment (targos) [inline src/util/getShapeGenerator.ts:9]:
```suggestion
type ShapeKind = 'gaussian' | 'gaussian2D' | 'lorentzian' | 'pseudoVoigt';
```

File: src/util/getShapeGenerator.ts [hunk 7]
```diff
@@ -3,8 +3,15 @@ import { Gaussian2D } from '../classes/Gaussian2D';
...
+
+/**
+ * Generate a instance of a specific kind of shape.
+ */
+export function getShapeGenerator(kind: shapeKind, shapeOptions = {}) {
   switch (kind.toLowerCase().replace(/[^a-z^0-9]/g, '')) {
```
Comment (targos) [inline src/util/getShapeGenerator.ts:15]:
```suggestion
  switch (kind) {
```

File: src/classes/Lorentzian.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+ * @param [area=0.9999] Expected area to be covered.
+ * @returns
+ */
+export function getFactor(area = 0.9999) {
+  return 2 * Math.tan(Math.PI * (area - 0.5));
+}
+
+/**
+ * Calculate intensity array of a lorentzian shape.
+ * @return {Float64Array} y values
```
Comment (lpatiny) [inline src/classes/Lorentzian.ts:146]: We could use everywhere @returns even if both are allowed

File: src/classes/__tests__/Gaussian.functions.test.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    let lenG = data.length;
+    let center = Math.floor((lenG - 1) / 2);
+    expect(data[center]).toBeCloseTo(1, 4);
+    expect(data[center - 1]).toBeCloseTo(data[center + 1], 4);
+    expect(data[center]).toBeGreaterThan(data[center + 1]);
+  });
+  it('even fwhm', () => {
+    let data = gaussian.getData({ length: 100, fwhm: 100, height: 1 });
+    expect(data).toHaveLength(100);
+    let lenG = data.length;
```
Comment (lpatiny) [inline src/classes/__tests__/Gaussian.functions.test.ts:58]: We could try to replace the `let` by `const` if it is constant values. There may be a lot of changes for this.

File: src/classes/__tests__/Gaussian.functions.test.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    // first derivative
+    yPrime[i] = y[i] - y[i - 1];
+  }
+
+  let positive = true;
+  let nbChanges = 0;
+  for (let i = 1; i < yPrime.length; i++) {
+    let diff = yPrime[i] - yPrime[i - 1];
+
+    if (diff > 0 && positive === false) {
```
Comment (lpatiny) [inline src/classes/__tests__/Gaussian.functions.test.ts:100]:
```suggestion
    if (diff > 0 && ! positive) {
```

File: src/classes/Gaussian2D.ts (hunk 5 above)
Comment (jobo322) [reply src/classes/Gaussian2D.ts:9]: it is needed when you point to properties inside of the object.

Comment (targos) [reply src/classes/Gaussian2D.ts:9]: can you show what breaks if you remove this line?

**Pull request merged**